# Modified to be able to :
#    - change the slave ID dynamically
#    - access new transaction counts for monitoring quality of connection
#    - made virtual methods that may be overridden by subclasses
diff --git a/src/ModbusMaster.cpp b/src/ModbusMaster.cpp
index 4169e58..eb4a943 100644
--- a/src/ModbusMaster.cpp
+++ b/src/ModbusMaster.cpp
@@ -46,6 +46,22 @@ ModbusMaster::ModbusMaster(void)
   _idle = 0;
   _preTransmission = 0;
   _postTransmission = 0;
+
+  _u32TotalTransactions = 0;
+  _u32FailedTransactions = 0;
+}
+
+void ModbusMaster::getTransactionCounts( uint32_t *total, uint32_t *failed )
+{
+   if ( total )
+   {
+      *total = _u32TotalTransactions;
+   }
+
+   if ( failed )
+   {
+      *failed = _u32FailedTransactions;
+   }
 }
 
 /**
@@ -72,6 +88,10 @@ void ModbusMaster::begin(uint8_t slave, Stream &serial)
 #endif
 }
 
+void ModbusMaster::setSlaveId(uint8_t slave)
+{
+   _u8MBSlave = slave;
+}
 
 void ModbusMaster::beginTransmission(uint16_t u16Address)
 {
@@ -872,5 +892,12 @@ uint8_t ModbusMaster::ModbusMasterTransaction(uint8_t u8MBFunction)
   _u8TransmitBufferIndex = 0;
   u16TransmitBufferLength = 0;
   _u8ResponseBufferIndex = 0;
+
+  _u32TotalTransactions++;
+  if ( u8MBStatus != ku8MBSuccess )
+  {
+   _u32FailedTransactions++;
+  }
+
   return u8MBStatus;
 }
diff --git a/src/ModbusMaster.h b/src/ModbusMaster.h
index 8c433e6..9e1b4d7 100644
--- a/src/ModbusMaster.h
+++ b/src/ModbusMaster.h
@@ -71,10 +71,12 @@ class ModbusMaster
   public:
     ModbusMaster();
    
-    void begin(uint8_t, Stream &serial);
-    void idle(void (*)());
-    void preTransmission(void (*)());
-    void postTransmission(void (*)());
+    virtual void begin(uint8_t, Stream &serial);
+    virtual void setSlaveId(uint8_t slave);
+    virtual void idle(void (*)());
+    virtual void preTransmission(void (*)());
+    virtual void postTransmission(void (*)());
+    virtual void getTransactionCounts( uint32_t *total, uint32_t *failed );
 
     // Modbus exception codes
     /**
@@ -187,37 +189,37 @@ class ModbusMaster
     @ingroup constant
     */
     static const uint8_t ku8MBInvalidCRC                 = 0xE3;
-    
-    uint16_t getResponseBuffer(uint8_t);
-    void     clearResponseBuffer();
-    uint8_t  setTransmitBuffer(uint8_t, uint16_t);
-    void     clearTransmitBuffer();
-    
-    void beginTransmission(uint16_t);
-    uint8_t requestFrom(uint16_t, uint16_t);
-    void sendBit(bool);
-    void send(uint8_t);
-    void send(uint16_t);
-    void send(uint32_t);
-    uint8_t available(void);
-    uint16_t receive(void);
-    
-    
-    uint8_t  readCoils(uint16_t, uint16_t);
-    uint8_t  readDiscreteInputs(uint16_t, uint16_t);
-    uint8_t  readHoldingRegisters(uint16_t, uint16_t);
-    uint8_t  readInputRegisters(uint16_t, uint8_t);
-    uint8_t  writeSingleCoil(uint16_t, uint8_t);
-    uint8_t  writeSingleRegister(uint16_t, uint16_t);
-    uint8_t  writeMultipleCoils(uint16_t, uint16_t);
-    uint8_t  writeMultipleCoils();
-    uint8_t  writeMultipleRegisters(uint16_t, uint16_t);
-    uint8_t  writeMultipleRegisters();
-    uint8_t  maskWriteRegister(uint16_t, uint16_t, uint16_t);
-    uint8_t  readWriteMultipleRegisters(uint16_t, uint16_t, uint16_t, uint16_t);
-    uint8_t  readWriteMultipleRegisters(uint16_t, uint16_t);
-    
-  private:
+
+    virtual uint16_t getResponseBuffer(uint8_t);
+    virtual void     clearResponseBuffer();
+    virtual uint8_t  setTransmitBuffer(uint8_t, uint16_t);
+    virtual void     clearTransmitBuffer();
+
+    virtual void beginTransmission(uint16_t);
+    virtual uint8_t requestFrom(uint16_t, uint16_t);
+    virtual void sendBit(bool);
+    virtual void send(uint8_t);
+    virtual void send(uint16_t);
+    virtual void send(uint32_t);
+    virtual uint8_t available(void);
+    virtual uint16_t receive(void);
+
+
+    virtual uint8_t  readCoils(uint16_t, uint16_t);
+    virtual uint8_t  readDiscreteInputs(uint16_t, uint16_t);
+    virtual uint8_t  readHoldingRegisters(uint16_t, uint16_t);
+    virtual uint8_t  readInputRegisters(uint16_t, uint8_t);
+    virtual uint8_t  writeSingleCoil(uint16_t, uint8_t);
+    virtual uint8_t  writeSingleRegister(uint16_t, uint16_t);
+    virtual uint8_t  writeMultipleCoils(uint16_t, uint16_t);
+    virtual uint8_t  writeMultipleCoils();
+    virtual uint8_t  writeMultipleRegisters(uint16_t, uint16_t);
+    virtual uint8_t  writeMultipleRegisters();
+    virtual uint8_t  maskWriteRegister(uint16_t, uint16_t, uint16_t);
+    virtual uint8_t  readWriteMultipleRegisters(uint16_t, uint16_t, uint16_t, uint16_t);
+    virtual uint8_t  readWriteMultipleRegisters(uint16_t, uint16_t);
+
+  protected:
     Stream* _serial;                                             ///< reference to serial port object
     uint8_t  _u8MBSlave;                                         ///< Modbus slave (1..255) initialized in begin()
     static const uint8_t ku8MaxBufferSize                = 64;   ///< size of response/transmit buffers    
@@ -233,7 +235,10 @@ class ModbusMaster
     uint16_t* rxBuffer; // from Wire.h -- need to clean this up Rx
     uint8_t _u8ResponseBufferIndex;
     uint8_t _u8ResponseBufferLength;
-    
+    uint32_t _u32TotalTransactions;                              ///< Number of transactions
+    uint32_t _u32FailedTransactions;                             ///< Failed transactions
+
+
     // Modbus function codes for bit access
     static const uint8_t ku8MBReadCoils                  = 0x01; ///< Modbus function 0x01 Read Coils
     static const uint8_t ku8MBReadDiscreteInputs         = 0x02; ///< Modbus function 0x02 Read Discrete Inputs
